File: Data Structures & Algorithms/course-schedule/submission-2.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        nodes = [Node(i) for i in range(numCourses)]
        roots = set([i for i in range(numCourses)])
        for pre in prerequisites:
            before, after = pre[1], pre[0]
            if before == after:
                return False
            roots.discard(after)
            nodes[before].children.append(nodes[after])

        def search(node):
            nonlocal triversed
            if node.val in triversed:
                return False
            triversed.add(node.val)
            for child in node.children:
                if search(child) == False:
                    return False
            triversed.remove(node.val)
            return True
        
        if len(roots) == 0:
            return False
        for r in roots:
            root = nodes[r]
            triversed = set()
            if search(root) == False:
                return False
        return True


class Node:
    def __init__(self, val):
        self.val = val
        self.children = []

File: Data Structures & Algorithms/n-ary-tree-postorder-traversal/submission-2.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> ret;

    void traverse(Node* node) {
        for (auto child : node->children) {
            traverse(child);
        }
        ret.push_back(node->val);
    }

    vector<int> postorder(Node* root) {
        if (root == nullptr) {
            return {};
        }
        traverse(root);
        return ret;
    }
};